cmd/argocd/commands/plugin.go

```diff
@@ -0,0 +1,171 @@
+		// The PluginPath is important to be returned since it
+		// helps us understanding the logic for handling errors.
+		log.Println("command does not exist, looking for a plugin...")
+		pluginPath, pluginErr := h.handlePluginCommand(args[1:], 1)
+		// IMP: If a plugin doesn't exist, the returned path will be empty along with nil error
+		// This means the command is neither a normal Argo CD Command nor a plugin.
+		if pluginErr == nil {
+			if pluginPath == "" {
+				log.Errorf("Error: %v\nRun 'argocd --help' for usage.\n", err)
+				return err
+			}
+		} else {
+			// If plugin handling fails, report the plugin error and exit
+			log.Errorf("Error: %v\n", pluginErr)
+			return pluginErr
+		}
+	} else {
+		// If it's any other error (not an unknown command), report it directly and exit
+		log.Errorf("Error: %v\n", err)
+		return err
+	}
+
+	return nil
+}
+
+// handlePluginCommand is  responsible for finding and executing a plugin when a command isn't recognized as a built-in command
+func (h *DefaultPluginHandler) handlePluginCommand(cmdArgs []string, minArgs int) (string, error) {
+	var remainingArgs []string // this will contain all "non-flag" arguments
+	foundPluginPath := ""
+	for _, arg := range cmdArgs {
+		// if you encounter a flag, break the loop
+		// For eg. If cmdArgs is ["argocd", "foo", "-v"],
+		// it will store ["argocd", "foo"] in remainingArgs
+		// and stop when it hits the flag -v
+		if strings.HasPrefix(arg, "-") {
+			break
+		}
+		remainingArgs = append(remainingArgs, strings.ReplaceAll(arg, "-", "_"))
+	}
+
+	if len(remainingArgs) == 0 {
+		// the length of cmdArgs is at least 1
+		return foundPluginPath, fmt.Errorf("flags cannot be placed before plugin name: %s", cmdArgs[0])
+	}
+
+	// try to find the binary, starting at longest possible name with given cmdArgs
+	for len(remainingArgs) > 0 {
+		path, found := h.lookForPlugin(strings.Join(remainingArgs, "-"))
+		if !found {
+			remainingArgs = remainingArgs[:len(remainingArgs)-1]
+			if len(remainingArgs) < minArgs {
+				break
+			}
+
+			continue
+		}
+
+		foundPluginPath = path
+		break
+	}
+
+	if len(foundPluginPath) == 0 {
+		return foundPluginPath, nil
+	}
+
+	// Execute the plugin that is found
+	if err := h.executePlugin(foundPluginPath, cmdArgs[len(remainingArgs):], os.Environ()); err != nil {
+		return foundPluginPath, err
+	}
+
+	return foundPluginPath, nil
+}
+
+// LookForPlugin implements PluginHandler. searches for an executable plugin with a given filename
+// and valid prefixes. If the plugin is not found (ErrNotFound), it continues to the next prefix.
+// Unexpected errors (e.g., permission issues) are logged for debugging but do not stop the search.
+// This doesn't care about the plugin execution errors since those errors are handled separately by
+// the execute function. Only those binaries are considered as argocd plugins if they start with
+// argocd and have executable permisssions. If they don't have executable permissions inspite of being
```

Comment:
> If they don't have executable permissions

I don't see the executable validation logic as part of this function. Please just document the function specific behaviour and move this doc to where the validation is actually implemented.

Reply:
the executable validation logic is different from what has been implemented here. `lookPath()` has a default value of `exec.lookPath()` that would find a plugin only if the following conditions are verfied:

1. A binary starting with `argocd` as the prefix.
2. The binary must have executable permissions.

Hence, if a binary doesn't satisfy these above conditions, lookPath() will give us an error and the plugin execution code never gets executed (why? because plugin is not found). Hence, this is the reason why the above comment you mentioned was documented for this function.